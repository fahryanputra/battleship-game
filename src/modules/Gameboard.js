import Ship from "../modules/Ship";

class Gameboard {
  // Gameboard legends
  // 0 = water
  // 1 = miss
  // 2 = hit

  // Gameboard axis
  // 0 = horizontal
  // 1 = vertical

  constructor(x, y) {
    this.x = x;
    this.y = y;
    this.message = "";
    this.ships = [];
    this.availableShips = [];
    this.board = [];

    this.createBoard();
  }

  createBoard() {
    for (let i = 0; i < this.x; i++) {
      this.board[i] = [];
      for (let j = 0; j < this.y; j++) {
        this.board[i].push(0);
      }
    }
  }

  addShip(name, length) {
    const ship = new Ship(name, length);
    this.availableShips.push(ship);
    this.ships.push(ship);
  }

  placeShip(name, x, y, axis) {
    const ship = this.ships.find((element) => element.getName() === name);

    if (axis === 0) {
      for (let i = y; i < ship.length; i++) {
        this.board[x][i] = ship;
      }
    } else {
      for (let i = x; i < ship.length; i++) {
        this.board[i][y] = ship;
      }
    }
  }

  destroyIfSunk(tile) {
    if (tile.getIsSunk() === true) {
      this.availableShips = this.availableShips.filter(
        (ship) => ship.getName() !== tile.getName(),
      );
    }
  }

  receiveAttack(x, y) {
    switch (this.board[x][y]) {
      case 0:
        this.board[x][y] = 1;
        this.message = "miss";
        break;

      case 1:
      case 2:
        this.message = "invalid";
        break;

      default:
        this.board[x][y].hit();
        this.destroyIfSunk(this.board[x][y]);
        this.board[x][y] = 2;
        this.message = "hit";
        break;
    }
  }

  // Getter
  getX() {
    return this.x;
  }

  getY() {
    return this.y;
  }

  getBoard() {
    return this.board;
  }

  getShips() {
    return this.ships;
  }

  getAvailableShips() {
    return this.availableShips;
  }

  getMessage() {
    return this.message;
  }
}

export default Gameboard;
