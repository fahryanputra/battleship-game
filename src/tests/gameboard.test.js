import { describe, expect, test } from "@jest/globals";
import Gameboard from "../modules/Gameboard";

const gameboard = new Gameboard(3, 3);

describe("Gameboard class", () => {
  test("create board", () => {
    expect(gameboard.getBoard()).toStrictEqual([
      [".", ".", "."],
      [".", ".", "."],
      [".", ".", "."],
    ]);
  });

  test("place ship", () => {
    gameboard.addShip("ship", 2, 0, 0, 0);
    expect(gameboard.getBoard()).toStrictEqual([
      ["ship", "ship", "."],
      [".", ".", "."],
      [".", ".", "."],
    ]);
  });
});
