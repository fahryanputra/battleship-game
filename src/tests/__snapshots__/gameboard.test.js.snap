// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gameboard class check ship object 1`] = `
{
  "damage": 0,
  "isSunk": false,
  "length": 2,
  "name": "destroyer",
}
`;

exports[`Gameboard class place another ship 1`] = `
{
  "damage": 0,
  "isSunk": false,
  "length": 3,
  "name": "cruiser",
}
`;

exports[`Gameboard class place another ship 2`] = `
{
  "damage": 0,
  "isSunk": false,
  "length": 3,
  "name": "cruiser",
}
`;

exports[`Gameboard class place another ship 3`] = `
{
  "damage": 0,
  "isSunk": false,
  "length": 3,
  "name": "cruiser",
}
`;

exports[`Gameboard class place ship 1`] = `
{
  "damage": 0,
  "isSunk": false,
  "length": 2,
  "name": "destroyer",
}
`;

exports[`Gameboard class place ship 2`] = `
{
  "damage": 0,
  "isSunk": false,
  "length": 2,
  "name": "destroyer",
}
`;
